### PR TITLE
Follow CNAME and DNAME records

### PR DESCRIPTION
**Pull request by mreiger**

### CODE REVIEW

**Comment by mreiger** [comment]
Hi @GrigoriyMikhalkin can you have a look at what I am trying here and tell me if this is a valid approach? (Following the CNAME redirection with recursive DNS lookups)

Im still a bit concerned if this works as intended; in a test I got a name with multiple CNAME redirections resolved even though there were more redirection levels than the TTL I built into the recursive lookup.
Also the code is at least somewhat ineffective: Usually the resolution of CNAME to A record comes in the same reply but this is currently ignored and a recursive DNS query vor the CNAME is done anyway.

So @chbmuc: I'd like to resolve these issues before I declare it ready for review.

**Comment by mreiger** [comment]
I think this is ready to look at now. Couple thoughts though:

* We probably should turn the logging down to a reasonable amount before releasing this though
* I wonder if we could use a context to break the recursion once the redirection TTL has been reached; but I'm not sure where to get the context from.

File: pkg/dns/dns_proxy_handler.go [hunk 1]
```diff
@@ -109,7 +110,9 @@ func getUpdateCacheFunc(log logr.Logger, cache *DNSCache) func(lookupTime time.T
 	return func(lookupTime time.Time, response *dnsgo.Msg) {
 		if response.Response && response.Rcode == dnsgo.RcodeSuccess {
 			scopedLog := log.WithValues(reqIdLogField, response.Id)
-			if err := cache.Update(lookupTime, response); err != nil {
+			qname := strings.ToLower(response.Question[0].Name)
```
Comment (majst01) [inline pkg/dns/dns_proxy_handler.go:113]: must be checked e.g. `if len(response.Question) > 0`

File: pkg/dns/dns_proxy_handler.go [hunk 2]
```diff
@@ -109,7 +110,9 @@ func getUpdateCacheFunc(log logr.Logger, cache *DNSCache) func(lookupTime time.T
...
+			log.Info("DEBUG dnsproxyhandler function getUpdateCacheFunc updating DNS cache", "queried name", qname, "dns response", response)
```
Comment (majst01) [inline pkg/dns/dns_proxy_handler.go:114]: For debug logging you can raise the verbosity like this: `log.V(4).Info("...")`. Default info level is V(0), V(4) is mostly used for debugging. See: https://pkg.go.dev/github.com/go-logr/logr#hdr-Verbosity

**Review by majst01 — CHANGES_REQUESTED** [review]

File: pkg/dns/dns_proxy_handler.go (hunk 1 above)
Comment (mreiger) [reply pkg/dns/dns_proxy_handler.go]: Should be fixed now; can you check if this is right?

File: pkg/dns/dns_proxy_handler.go (hunk 2 above)
Comment (mreiger) [reply pkg/dns/dns_proxy_handler.go]: Done!

**Review by mwennrich — APPROVED** [review]

**Review by Gerrit91 — APPROVED** [review]

**Review by majst01 — APPROVED** [review]

**Pull request merged**